cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AndTrigger.java

```diff
@@ -46,4 +49,32 @@ public boolean isSatisfied(ProgramSchedule schedule, List<Notification> notifica
   public List<TriggerInfo> getTriggerInfos(TriggerInfoContext context) {
     return getUnitTriggerInfosAddRuntimeArgs(context);
   }
+
+  @Nullable
+  @Override
+  public AbstractCompositeTrigger getTriggerWithDeletedProgram(ProgramId programId) {
+    List<SatisfiableTrigger> updatedTriggers = new ArrayList<>();
+    for (SatisfiableTrigger trigger : getSatisfiableTriggers()) {
+      if (trigger instanceof ProgramStatusTrigger &&
+        programId.equals(((ProgramStatusTrigger) trigger).getProgramId())) {
+        // this program status trigger will never be satisfied, so the current AND trigger will never be satisfied
+        return null;
+      }
+      if (trigger instanceof AbstractCompositeTrigger) {
+        SatisfiableTrigger updatedTrigger = ((AbstractCompositeTrigger) trigger)
+          .getTriggerWithDeletedProgram(programId);
+        if (updatedTrigger == null) {
+          // the updated composite trigger will never be satisfied, so the AND trigger will never be satisfied
+          return null;
+        }
+        // add the updated composite trigger into updatedTriggers
+        updatedTriggers.add(updatedTrigger);
+      } else {
+        // the trigger is not a composite trigger, add it to updatedTriggers directly
+        updatedTriggers.add(trigger);
+      }
+    }
+    // return a new AND trigger constructed from the updated triggers
+    return new AndTrigger(updatedTriggers.toArray(new SatisfiableTrigger[updatedTriggers.size()]));
```

Comment:
why does the constructor take an array and not a list?

Reply:
especially since it looks like the array is just eventually turned into a list.

Reply:
seems like if updatedTriggers is size 1, you should just return the trigger instead of making an And of one thing.